File: MidtermLab1/JavaScript/product_category_profitability_analyzer.js
const fs = require('fs');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askFilePath() {
    rl.question("Enter dataset file path (vgchartz-2024.csv): ", (path) => {
        if (fs.existsSync(path) && fs.lstatSync(path).isFile()) {
            processCSV(path);
        } else {
            console.log("Invalid file path. Please try again.");
            askFilePath();
        }
    });
}

function processCSV(path) {
    const data = fs.readFileSync(path, 'utf8');
    const lines = data.split('\n');
    
    const salesMap = {};
    const countMap = {};

    // Start from 1 to skip header
    for (let i = 1; i < lines.length; i++) {
        const line = lines[i].trim();
        if (!line) continue;

        // Regex to handle commas inside quotes
        const columns = line.split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/);

        if (columns.length >= 9) {
            const category = columns[3].trim(); // Genre
            const sales = parseFloat(columns[7].trim()) || 0; // Total Sales

            salesMap[category] = (salesMap[category] || 0) + sales;
            countMap[category] = (countMap[category] || 0) + 1;
        }
    }

    displayResults(salesMap, countMap);
}

function displayResults(sales, counts) {
    let mostProfitable = "";
    let leastProfitable = "";
    let maxS = -1;
    let minS = Infinity;

    console.log("\n--- Category Analytics ---");
    for (const cat in sales) {
        const total = sales[cat];
        const avg = total / counts[cat];

        console.log(`Category: ${cat.padEnd(15)} | Total: ${total.toFixed(2).padStart(10)} | Avg: ${avg.toFixed(2).padStart(8)}`);

        if (total > maxS) {
            maxS = total;
            mostProfitable = cat;
        }
        if (total < minS) {
            minS = total;
            leastProfitable = cat;
        }
    }

    console.log(`\nMost Profitable Category: ${mostProfitable}`);
    console.log(`Least Profitable Category: ${leastProfitable}`);
    rl.close();
}

askFilePath();

File: MidtermLab1/ProductCategoryProfitabilityAnalyzer/ProductCategoryProfitabilityAnalyzer.java
import java.util.*;
import java.io.*;

public class ProductCategoryProfitabilityAnalyzer {
    public static void main(String[] args) {
        //  Requirement: Use Scanner for input
        Scanner input = new Scanner(System.in); 
        File file;

        //  Requirement: Loop until valid path
        while (true) {
            System.out.print("Enter dataset file path (vgchartz-2024.csv): ");
            String path = input.nextLine();
            file = new File(path); //  Requirement: Use File class for validation

            if (file.exists() && file.isFile()) {
                break; 
            } else {
                System.out.println("Invalid file path. Please try again.");
            }
        }

        Map<String, Double> totalSalesMap = new HashMap<>();
        Map<String, Integer> countMap = new HashMap<>();

        // Requirement: Load dataset into memory
        try (Scanner reader = new Scanner(file)) {
            if (reader.hasNextLine()) reader.nextLine(); // Skip CSV Header

            while (reader.hasNextLine()) {
                String line = reader.nextLine();
                // Regex handles commas inside quotes (common in game titles)
                String[] data = line.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
                
                if (data.length < 9) continue; 

                // Requirement: Group by product category (Genre)
                String category = data[3].trim(); 
                try {
                    // Using Column 8 for Total Sales from vgchartz-2024.csv
                    double sales = Double.parseDouble(data[8].trim()); 
                    
                    totalSalesMap.put(category, totalSalesMap.getOrDefault(category, 0.0) + sales);
                    countMap.put(category, countMap.getOrDefault(category, 0) + 1);
                } catch (NumberFormatException e) {
                    continue; 
                }
            }

            // Requirement: Perform analytics and display formatted results
            displayResults(totalSalesMap, countMap); 
        } catch (Exception e) {
            System.out.println("Error processing data: " + e.getMessage());
        } finally {
            // FIX: Scanner closed here to resolve the "resource leak" issue
            input.close(); 
        }
    }

    private static void displayResults(Map<String, Double> sales, Map<String, Integer> counts) {
        String mostProfitable = "";
        String leastProfitable = "";
        double maxSales = -1;
        double minSales = Double.MAX_VALUE;

        System.out.println("\n--- Category Analytics ---");
        for (String cat : sales.keySet()) {
            double total = sales.get(cat); // Requirement: Total sales per category
            double avg = total / counts.get(cat); // Requirement: Average sale per category
            
            System.out.printf("Category: %-15s | Total: %10.2f | Avg: %8.2f\n", cat, total, avg);

            // Identify Most and Least profitable
            if (total > maxSales) {
                maxSales = total;
                mostProfitable = cat;
            }
            if (total < minSales) {
                minSales = total;
                leastProfitable = cat;
            }
        }
        System.out.println("\nMost Profitable Category: " + mostProfitable);
        System.out.println("Least Profitable Category: " + leastProfitable);
    }
}